engine/consensus/matching/engine.go

```diff
@@ -651,37 +703,75 @@ func (e *Engine) sealableResults() ([]*flow.IncorporatedResult, error) {
 		// when we seal ERs that don't have enough approvals but are deep enough in the chain resulting in halting sealing
 		// process. This will be removed when implementation of seal & verification is finished.
 		if !matched && e.emergencySealingActive {
-			block, err := e.headersDB.ByBlockID(incorporatedResult.IncorporatedBlockID)
+			incorporatedBlock, err := e.headersDB.ByBlockID(incorporatedResult.IncorporatedBlockID)
 			if err != nil {
-				return nil, fmt.Errorf("could not get block %v: %w", incorporatedResult.IncorporatedBlockID, err)
+				return nil, nil, fmt.Errorf("could not get block %v: %w", incorporatedResult.IncorporatedBlockID, err)
 			}
-			if block.Height+DefaultEmergencySealingThreshold <= lastFinalized.Height {
-				matched = true
+			// Note:
+			// we assume the incorporatedBlock is for a unsealed block, if
+			// there are DefaultEmergencySealingThreshold number of blocks between incorporatedBlock
+			// and lastFinalized, it means there are as many unsealed and finalized blocks, because
+			// should trigger the emergency sealing
+			if incorporatedBlock.Height+DefaultEmergencySealingThreshold <= lastFinalized.Height {
+				matchedAfterConsideringEmergency = true
 			}
 		}
 
-		if matched {
+		if matchedAfterConsideringEmergency {
 			// add the result to the results that should be sealed
 			results = append(results, incorporatedResult)
 		}
+
+		// whether we sealed the result when the emergency sealing is active and triggered
+		sealedByEmergency := !matched && matchedAfterConsideringEmergency
+
+		if nextUnsealedIsFinalized {
+			if incorporatedResult.Result.BlockID == nextUnsealed {
+				nextUnsealeds = append(nextUnsealeds, &nextUnsealedResult{
+					BlockID:                  incorporatedResult.Result.BlockID,
+					Height:                   block.Height,
+					ResultID:                 incorporatedResult.Result.ID(),
+					IncorporatedResultID:     incorporatedResult.ID(),
+					TotalChunks:              len(incorporatedResult.Result.Chunks),
+					FirstUnmatchedChunkIndex: unmatchedIndex,
+					CanBeSealed:              matchedAfterConsideringEmergency,
+					SealedByEmergency:        sealedByEmergency,
+				})
+			}
+		}
 	}
 
-	return results, nil
+	return results, nextUnsealeds, nil
+}
+
+// This struct is made for debugging potential sealing halt
+type nextUnsealedResult struct {
+	BlockID                  flow.Identifier // the block of of the next unsealed block
+	Height                   uint64          // the height of the block
+	ResultID                 flow.Identifier // if we haven't received the result, it would be ZeroID
+	IncorporatedResultID     flow.Identifier // to find seal in mempool
+	TotalChunks              int
+	FirstUnmatchedChunkIndex int  // show which chunk hasn't received approval
+	CanBeSealed              bool // if true, then it should soon go to seals mempool
+	SealedByEmergency        bool // if sealed by emergency since there are too many unsealed blocks
```

Comment:
how about we move this to its own file, or to the end of the engine implementation. To me, it seems a little out of place here in the middle between all the engine's methods